wgpu-core/src/device/mod.rs

```diff
@@ -2575,3 +2618,30 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         buffer.map_state = resource::BufferMapState::Idle;
     }
 }
+
+fn validate_shader(module: &naga::Module, entry_point_name: &str, execution_model: ExecutionModel) {
+    let entry_point = module
+        .entry_points
+        .iter()
+        .find(|entry_point| entry_point_name == entry_point.name);
+
+    let entry_point = match entry_point {
+        Some(entry_point) => entry_point,
+        None => {
+            log::error!(
+                "Unable to find entry point `{}` in shader module",
+                entry_point_name
+            );
+            return;
+        }
+    };
+
+    if entry_point.exec_model != execution_model {
```

Comment:
Note that it's legal for a module to have multiple entry points named the same if they are for different stages. So there needs to be only one loop over them, comparing both the name and the exec model. Consequently, it will be only one kind of an error for this (e.g. `MissingEntryPoint`).